### PR TITLE
Fix build for android

### PR DESCRIPTION
**Pull request by ephemer**
As mentioned in #312, this fixes the build for Android and works on both Mac and Linux with the following CMake invocation:

```
cmake -G Ninja \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_SYSTEM_NAME=Android \
    -DCMAKE_SYSROOT=$ANDROID_STANDALONE_TOOLCHAIN/sysroot \
    -DENABLE_SWIFT=YES \
    -DBUILD_SHARED_LIBS=YES \
    -DCMAKE_SWIFT_COMPILER=$SWIFT_INSTALLATION_PATH/usr/Darwin/swiftc \
    -DENABLE_TESTING=OFF \
    -DCMAKE_ANDROID_ARCH_ABI=armeabi-v7a \
    -DCMAKE_ANDROID_NDK_TOOLCHAIN_VERSION=clang \
    $SWIFT_SOURCE_PATH/swift-corelibs-libdispatch
```

### CODE REVIEW

File: dispatch/dispatch.h [hunk 1]
```diff
@@ -30,6 +30,8 @@
 #include <os/linux_base.h>
 #endif
 
+#include <string.h>
```
Comment (ephemer) [inline dispatch/dispatch.h]: This should probably be wrapped with an #if ANDROID

File: cmake/modules/SwiftSupport.cmake [hunk 2]
```diff
@@ -38,6 +38,10 @@ function(add_swift_library library)
     endforeach()
   endif()
 
+  if(CMAKE_SYSTEM_NAME STREQUAL Android)
+    list(APPEND flags -sdk;${CMAKE_SYSROOT})
```
Comment (ephemer) [inline cmake/modules/SwiftSupport.cmake]: It probably makes sense to always add the sysroot?

File: cmake/modules/SwiftSupport.cmake [hunk 3]
```diff
@@ -13,7 +13,7 @@ function(add_swift_library library)
   list(APPEND flags -emit-library)
 
   if(ASL_TARGET)
-    list(APPEND FLAGS -target;${ASL_TARGET})
+    list(APPEND flags -target;${ASL_TARGET})
```
Comment (ephemer) [inline cmake/modules/SwiftSupport.cmake]: apparently CMake variable names are case sensitive. This is a typo introduced in an earlier PR

**Comment by MadCoder** [comment]
@compnerd you did most of the cmake porting, can you review the change please?

File: dispatch/dispatch.h [hunk 4]
```diff
@@ -30,6 +30,10 @@
 #include <os/linux_base.h>
 #endif
 
+#ifdef __ANDROID__
+#include <string.h>
```
Comment (compnerd) [inline dispatch/dispatch.h]: I think that this can be done unconditionally.

File: cmake/modules/SwiftSupport.cmake (hunk 2 above)
Comment (compnerd) [reply cmake/modules/SwiftSupport.cmake]: Yeah, I think that we should always add the sysroot here.

**Comment by compnerd** [comment]
Can you please squash the changes into a single commit?

File: dispatch/dispatch.h [hunk 5]
```diff
@@ -30,6 +30,8 @@
 #include <os/linux_base.h>
 #endif
 
+
```
Comment (MadCoder) [inline dispatch/dispatch.h]: please remove the empty line

File: cmake/modules/SwiftSupport.cmake [hunk 6]
```diff
@@ -93,7 +95,7 @@ function(get_swift_host_arch result_var_name)
     set("${result_var_name}" "s390x" PARENT_SCOPE)
   elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv6l")
     set("${result_var_name}" "armv6" PARENT_SCOPE)
-  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7l")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "^armv7")
```
Comment (MadCoder) [inline cmake/modules/SwiftSupport.cmake]: do we want the same for armv6l above?

File: src/shims/android_stubs.h [hunk 7]
```diff
@@ -20,4 +20,8 @@
 #define _Static_assert(...)
 #endif
 
+#ifndef TRASHIT
```
Comment (MadCoder) [inline src/shims/android_stubs.h]: I'm fairly sure this has been moved to `internal.h` already in a previous PR, please remove.

**Comment by MadCoder** [comment]
Also please squash in a single commit.

**Comment by Fischiii** [comment]
@MadCoder can you elaborate? We are successfully working with this fork, since we need the android support. Would be nice to have this merged.

**Comment by compnerd** [comment]
I'm going to close this PR for now.  It no longer cleanly applies.  I believe that libdispatch builds for android currently, but, if it does not, that is certainly that should be addressed.

**Pull request closed without merging**